feat: Add Kubernetes  auth provider to use SelfSubjectReview and kubernetes api server

akram opened the pull request:
# What does this PR do?
Add Kubernetes authentication provider support
- Add KubernetesAuthProvider class for token validation using Kubernetes SelfSubjectReview API
- Add KubernetesAuthProviderConfig with configurable API server URL, TLS settings, and claims mapping
- Implement authentication via POST requests to /apis/authentication.k8s.io/v1/selfsubjectreviews endpoint
- Add support for parsing Kubernetes SelfSubjectReview response format to extract user information
- Add KUBERNETES provider type to AuthProviderType enum
- Update create_auth_provider factory function to handle 'kubernetes' provider type
- Add comprehensive unit tests for KubernetesAuthProvider functionality
- Add documentation with configuration examples and usage instructions

The provider validates tokens by sending SelfSubjectReview requests to the Kubernetes API server and extracts user information from the userInfo structure in the response.


<!-- If resolving an issue, uncomment and update the line below -->
<!-- Closes #[issue-number] -->

## Test Plan
<!-- Describe the tests you ran to verify your changes with result summaries. *Provide clear instructions so the plan can be easily re-executed.* -->
What This Verifies:
Authentication header validation
Token validation with Kubernetes SelfSubjectReview and kubernetes server API endpoint
Error handling for invalid tokens and HTTP errors
Request payload structure and headers

```
python -m pytest tests/unit/server/test_auth.py -k "kubernetes" -v
```

akram commented:
@leseb can you PTAL ?

ChristianZaccaria reviewed (commented):
ignore this comment, wrong PR

rhuss commented on llama_stack/distribution/server/auth_providers.py:
```diff
@@ -339,6 +339,134 @@ async def close(self):
+                user_attributes = {}
+                
+                # Map username
+                if "username" in self.config.claims_mapping:
```
username is already verified to be non-nil at this point (see line 413)

akram commented on llama_stack/distribution/server/auth_providers.py:
```diff
@@ -339,6 +339,134 @@ async def close(self):
+            logger.exception("Error during token validation")
+            raise ValueError(f"Token validation error: {str(e)}") from e
+    
+    async def close(self):
+        """Close any resources."""
+        pass
```
It is the implementation of the abstract method defined in `AuthProvider`.

Even though currently there is no need of cleanup (  a new HTTP client is created per request), having the method available makes it easy to add cleanup logic later if needed. And, it is made mandatory by the `@abstractmethod`

grs commented on llama_stack/distribution/server/auth_providers.py:
```diff
@@ -339,6 +339,134 @@ async def close(self):
             self._client = None
 
 
+class OpenShiftOAuthAuthProviderConfig(BaseModel):
```
The 'OAuth' in the name here seems misleading to me as the actual implementation does not rely on any oauth related specification as far as I can see. OpenshiftAuthProviderConfig would be clearer.

Would the use case be met by using the standard kubernetes SelfSubjecReview[1]? Can openshift not issue jwt tokens for users?

[1] https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/self-subject-review-v1/

grs commented on llama_stack/distribution/server/auth_providers.py:
```diff
@@ -339,6 +339,134 @@ async def close(self):
+                if not username:
+                    raise ValueError("No username found in user response")
+
+                uid = metadata.get("uid")
```
Is the uid here a uuid as with other kubernetes objects? If so is this field actually useful in this context over the simpler username as retrieved above? Seems like it would be rather unwieldy to use.

akram replied:
That's correct, uid is not relevant here.

grs commented on llama_stack/distribution/server/auth_providers.py:
```diff
@@ -347,6 +475,8 @@ def create_auth_provider(config: AuthenticationConfig) -> AuthProvider:
         return CustomAuthProvider(CustomAuthProviderConfig.model_validate(config.config))
     elif provider_type == "oauth2_token":
         return OAuth2TokenAuthProvider(OAuth2TokenAuthProviderConfig.model_validate(config.config))
+    elif provider_type == "openshift_oauth":
```
Again, the 'oauth' here is in my view misleading. This is simply an openshift specific approach to token verification is it not?

akram replied:
that's correct. I can rename it.

akram commented:
@grs can you PTAL ?

akram commented on llama_stack/distribution/datatypes.py:
```diff
@@ -251,8 +252,35 @@ class GitHubTokenAuthConfig(BaseModel):
+    claims_mapping: dict[str, str] = Field(
+        default_factory=lambda: {
+            "username": "username",
+            "groups": "roles",
```
done

grs commented on docs/source/distributions/configuration.md:
````diff
@@ -342,6 +342,46 @@ server:
+```bash
+kubectl create namespace llama-stack
+kubectl create serviceaccount llama-stack-auth -n llama-stack
+kubectl create token llama-stack-auth -n llama-stack > llama-stack-auth-token
````
Just out of curiousity, is the token you get from this not a jwt token? (e.g. does this show anything readable: `cat llama-stack-auth-token |  cut -d. -f2 | base64 -d`)

leseb reviewed (commented):
I believe this is how it was originally implemented (more or less) and removed it https://github.com/meta-llama/llama-stack/pull/2216 in favor of OAUTH2_TOKEN and the JWT endpoint of Kubernetes.

What's the reason to re-add this and not stick with JWT?

akram commented:
> I believe this is how it was originally implemented (more or less) and removed it #2216 in favor of OAUTH2_TOKEN and the JWT endpoint of Kubernetes.
> 
> What's the reason to re-add this and not stick with JWT?

The reason behind is mostly OpenShift related. OpenShift Console tokens cannot be validated with JWT as they are actually Opaque token. Technically, it is more a has of token than a token itself, this is why proper JWT + JWKS cannot be used. When OpenShift Console is used as an oauth provider, the issued tokens cannot be validated with JWKS but they can be validated through SelfSubjectReview.

There are probably some cases where this provided can be used, for example if a fully compliant JWT infrastructure is not enabled and users still want to have token validation.

leseb commented on docs/source/distributions/configuration.md:
```diff
@@ -365,6 +365,46 @@ server:
+server:
+  auth:
+    provider_config:
+      type: "kubernetes"
+      api_server_url: https://kubernetes.default.svc
+      claims_mapping:
+        username: "roles"
+        groups: "roles"
+        uid: "uid_attr"
+      verify_tls: true
+      tls_cafile: "/path/to/ca.crt"
```
the surrounding double quotes can be removed, rigth?

leseb commented on llama_stack/core/datatypes.py:
```diff
@@ -282,8 +283,35 @@ class GitHubTokenAuthConfig(BaseModel):
+        description="Mapping of Kubernetes user claims to access attributes",
+    )
+
+    @field_validator("claims_mapping")
```
another validator with parseurl for api_server_url?

leseb commented on llama_stack/core/server/auth_providers.py:
```diff
@@ -374,6 +375,90 @@ async def _get_github_user_info(access_token: str, github_api_base_url: str) ->
+        review_request = {"apiVersion": "authentication.k8s.io/v1", "kind": "SelfSubjectReview"}
+
+        try:
+            async with httpx.AsyncClient(verify=ssl_ctxt if ssl_ctxt else self.config.verify_tls) as client:
```
don't we have the right content for ssl_ctxt already? coming from above? why another condition?

akram replied:
previously, there was 2 cases where we build the ssl_ctxt : 
- ssl verification is disabled
- custom ca certificate is passed

if none of them happened, we would get an None context ... 
changing this.

leseb commented on llama_stack/core/server/auth_providers.py:
```diff
@@ -374,6 +375,90 @@ async def _get_github_user_info(access_token: str, github_api_base_url: str) ->
+                    timeout=10.0,
+                )
+
+                if response.status_code == 401:
+                    raise ValueError("Invalid token")
+                elif response.status_code != 201:
+                    logger.warning(f"Kubernetes SelfSubjectReview API failed with status code: {response.status_code}")
+                    raise ValueError(f"Token validation failed: {response.status_code}")
```
no action required: given that each condition raise something we could avoid the if elif and just have ifs

leseb commented on llama_stack/core/server/auth_providers.py:
```diff
@@ -374,6 +375,90 @@ async def _get_github_user_info(access_token: str, github_api_base_url: str) ->
+        except httpx.TimeoutException:
+            logger.warning("Kubernetes SelfSubjectReview API request timed out")
+            raise ValueError("Token validation timeout") from None
+        except ValueError:
+            # Re-raise ValueError exceptions to preserve their message
+            raise
```
Why is this needed?  If you want ValueErrors to propagate unchanged, you don’t need to catch them at all.

leseb commented on docs/source/distributions/configuration.md:
````diff
@@ -365,6 +365,46 @@ server:
 ```
 
 The provider fetches user information from GitHub and maps it to access attributes based on the `claims_mapping` configuration.
+#### Kubernetes Authentication Provider
````
Maybe this should be right after the existing Kube config with OAUTH?

akram commented:
> a few more nits

fixed 
and I also fixed https://github.com/llamastack/llama-stack/issues/3368 for the 2 remaining `200` for the other auth providers.

leseb commented on llama_stack/core/server/auth_providers.py:
```diff
@@ -374,6 +376,89 @@ async def _get_github_user_info(access_token: str, github_api_base_url: str) ->
+
+        except httpx.TimeoutException:
+            logger.warning("Kubernetes SelfSubjectReview API request timed out")
+            raise ValueError("Token validation timeout") from None
```
```suggestion
            raise TokenValidationError("Token validation timeout") from None
```

leseb commented on llama_stack/core/server/auth_providers.py:
```diff
@@ -374,6 +376,89 @@ async def _get_github_user_info(access_token: str, github_api_base_url: str) ->
+            raise ValueError("Token validation timeout") from None
+        except Exception as e:
+            logger.warning(f"Error during token validation: {str(e)}")
+            raise ValueError(f"Token validation error: {str(e)}") from e
```
```suggestion
            raise TokenValidationError(f"Token validation error: {str(e)}") from e
```
